.github/workflows/type-check.yml

```diff
@@ -0,0 +1,61 @@
+name: "Types: check published types"
+
+on: [pull_request, push]
+
+permissions:
+  contents: read
+
+jobs:
+  test:
+    name: TS ${{ matrix.ts_version }}, "${{ matrix.ts_lib }}"
+    runs-on: ubuntu-latest
+    strategy:
+      fail-fast: false
+      matrix:
+        ts_version:
+          # The official ESLint types are not compatible with TS 3.9
+          # - 3.9
+          - '4.0'
+          - 4.1
+          - 4.2
+          - 4.3
+          - 4.4
+          - 4.5
+          - '5.0'
+          - 5.5
+          - 5.6
+        ts_lib:
+          - es2015
+          - es2015,dom
+          - es2020
+          - esnext
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          persist-credentials: false
+          show-progress: false
```

Comment:
why are these helpful?

Reply:
`persist-credentials` as otherwise the token used to access the git repo will be saved in the `.git` config, which has caused security issues when people have eg zipped the entire folder or through some other way unintentionally shared that config, not knowing that they shared a token with permissions to do stuff others shouldn't be able to do. For this specific workflow it's not really needed as the permissions are explicitly set to be only for content and only for read – but I typically add it anyway in case permissions are extended in the future. I really think this should have been the default value for checkout.

`show-progress` is just to get a cleaner log – the progress log is just noise. It came with my copy and paste from my own workflow

Reply:
Thanks - i assume that if this is set to false, then future steps can't do git operations?

Reply:
Yes, I believe that to be correct

It's added to allow one to later do a `git push`, but there are better ways to do that: https://github.com/neostandard/neostandard/blob/79da5b6283d7611347cadaad7dc342e6a9bd9338/.github/workflows/sync-reusable.yml#L125

Reply:
hm, if `gh auth setup-git` gives you the same credentials that `persist-credentials` would give, then it seems like the only protection it offers is in case someone accidentally tries to tar up a repo's `.git` directory - that seems like a very unlikely scenario.

Reply:
It's actually something that was discovered recently to have happened fairly widespread – and prompted GitHub to change some of their official actions to exclude hidden files by default, which they did in a minor release to try to stop the issue as soon as possible, which in turn broke a lot of CI setups that relied on hidden files being included

Reply:
The trick with `gh auth setup-git` is that it's only available there and not anywhere else – not for any install scripts in prior steps, not for any actions in prior steps – and since I don't execute any code there the attack vectors are very slim (essentially git hooks or if somehow the built in git commands would be shimmed or something)

Reply:
right, but it's just a shell command, any part of the CI workflow can invoke it, no?

Reply:
The token it uses is provided specifically for that step: https://github.com/neostandard/neostandard/blob/79da5b6283d7611347cadaad7dc342e6a9bd9338/.github/workflows/sync-reusable.yml#L130

Reply:
Here's an article on what I mentioned yesterday – didn't find it when I was on my phone: https://www.bleepingcomputer.com/news/security/github-actions-artifacts-found-leaking-auth-tokens-in-popular-projects/

Reply:
I agree with this conclusion in the end of that article:

> The default settings for dangerous actions like 'actions/checkout' should be adjusted so credentials do not persist. Furthermore, token permissions used in workflows should be set to the least privilege necessary to limit damage in case they are exposed.